src/compile/scale.ts

```diff
@@ -43,8 +51,46 @@ export function compileScales(channels: Channel[], model: Model) {
         }
       });
 
-      return scaleDef;
-    });
+      // add additional scales needed to support ordinal color scales also for legends
+      if (channel === COLOR && fieldDef.legend && (fieldDef.type === ORDINAL || fieldDef.bin || fieldDef.timeUnit)) {
+        scales.push({
+          name: INVERSE_RANK,
+          type: ORDINAL,
+          domain: {
+            data: model.dataTable(),
+            // use rank_<field> for ordinal type, for bin and timeUnit use default field
+            field: model.field(COLOR, (fieldDef.bin || fieldDef.timeUnit) ? {} : {prefn: 'rank_'}), sort: true
+          },
+          range: {data: model.dataTable(), field: model.field(COLOR), sort: true}
```

Comment:
This INVERSE_RANK scale for bin and timeUnit is not necessary because you do not have rank for these two cases so you do not have to invert. 

In fact, your current code for INVERSE_RANK produces identity scale (`f(x)=x`) for bin & timeUnit cases so I think you can simplify a lot of code here.